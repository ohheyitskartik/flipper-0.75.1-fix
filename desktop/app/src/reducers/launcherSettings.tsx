/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Actions} from './index';
import ReleaseChannel from '../ReleaseChannel';

export type LauncherSettings = {
  releaseChannel: ReleaseChannel;
  ignoreLocalPin: boolean;
};

export type Action = {
  type: 'UPDATE_LAUNCHER_SETTINGS';
  payload: LauncherSettings;
};

export const defaultLauncherSettings: LauncherSettings = {
  releaseChannel: ReleaseChannel.DEFAULT,
  ignoreLocalPin: false,
};

export default function reducer(
  state: LauncherSettings = defaultLauncherSettings,
  action: Actions,
): LauncherSettings {
  if (action.type === 'UPDATE_LAUNCHER_SETTINGS') {
    return action.payload;
  }
  return state;
}

export function updateLauncherSettings(settings: LauncherSettings): Action {
  return {
    type: 'UPDATE_LAUNCHER_SETTINGS',
    payload: settings,
  };
}
