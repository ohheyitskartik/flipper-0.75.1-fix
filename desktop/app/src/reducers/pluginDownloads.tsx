/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {
  DownloadablePluginDetails,
  getPluginVersionInstallationDir,
} from 'flipper-plugin-lib';
import {Actions} from '.';
import produce from 'immer';
import {Canceler} from 'axios';

export enum PluginDownloadStatus {
  QUEUED = 'Queued',
  STARTED = 'Started',
  FAILED = 'Failed',
}

export type DownloadablePluginState = {
  plugin: DownloadablePluginDetails;
  startedByUser: boolean;
} & (
  | {status: PluginDownloadStatus.QUEUED}
  | {status: PluginDownloadStatus.STARTED; cancel: Canceler}
);

// We use plugin installation path as key as it is unique for each plugin version.
export type State = Record<string, DownloadablePluginState>;

export type PluginDownloadStart = {
  type: 'PLUGIN_DOWNLOAD_START';
  payload: {
    plugin: DownloadablePluginDetails;
    startedByUser: boolean;
  };
};

export type PluginDownloadStarted = {
  type: 'PLUGIN_DOWNLOAD_STARTED';
  payload: {
    plugin: DownloadablePluginDetails;
    cancel: Canceler;
  };
};

export type PluginDownloadFinished = {
  type: 'PLUGIN_DOWNLOAD_FINISHED';
  payload: {
    plugin: DownloadablePluginDetails;
  };
};

export type Action =
  | PluginDownloadStart
  | PluginDownloadStarted
  | PluginDownloadFinished;

const INITIAL_STATE: State = {};

export default function reducer(
  state: State = INITIAL_STATE,
  action: Actions,
): State {
  switch (action.type) {
    case 'PLUGIN_DOWNLOAD_START': {
      const {plugin, startedByUser} = action.payload;
      const installationDir = getPluginVersionInstallationDir(
        plugin.name,
        plugin.version,
      );
      const downloadState = state[installationDir];
      if (downloadState) {
        // If download is already in progress - re-use the existing state.
        return produce(state, (draft) => {
          draft[installationDir] = {
            ...downloadState,
            startedByUser: startedByUser || downloadState.startedByUser,
          };
        });
      }
      return produce(state, (draft) => {
        draft[installationDir] = {
          plugin,
          startedByUser: startedByUser,
          status: PluginDownloadStatus.QUEUED,
        };
      });
    }
    case 'PLUGIN_DOWNLOAD_STARTED': {
      const {plugin, cancel} = action.payload;
      const installationDir = getPluginVersionInstallationDir(
        plugin.name,
        plugin.version,
      );
      const downloadState = state[installationDir];
      if (downloadState?.status !== PluginDownloadStatus.QUEUED) {
        console.warn(
          `Invalid state transition PLUGIN_DOWNLOAD_STARTED in status ${downloadState?.status} for download to directory ${installationDir}.`,
        );
        return state;
      }
      return produce(state, (draft) => {
        draft[installationDir] = {
          status: PluginDownloadStatus.STARTED,
          plugin,
          startedByUser: downloadState.startedByUser,
          cancel,
        };
      });
    }
    case 'PLUGIN_DOWNLOAD_FINISHED': {
      const {plugin} = action.payload;
      const installationDir = getPluginVersionInstallationDir(
        plugin.name,
        plugin.version,
      );
      return produce(state, (draft) => {
        delete draft[installationDir];
      });
    }
    default:
      return state;
  }
}

export const startPluginDownload = (payload: {
  plugin: DownloadablePluginDetails;
  startedByUser: boolean;
}): Action => ({
  type: 'PLUGIN_DOWNLOAD_START',
  payload,
});

export const pluginDownloadStarted = (payload: {
  plugin: DownloadablePluginDetails;
  cancel: Canceler;
}): Action => ({type: 'PLUGIN_DOWNLOAD_STARTED', payload});

export const pluginDownloadFinished = (payload: {
  plugin: DownloadablePluginDetails;
}): Action => ({type: 'PLUGIN_DOWNLOAD_FINISHED', payload});
