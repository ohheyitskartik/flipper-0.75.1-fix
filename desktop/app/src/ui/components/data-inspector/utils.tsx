/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import naturalCompare from 'string-natural-compare';

export function getSortedKeys(obj: Object): Array<string> {
  return Object.keys(obj).sort(naturalCompare);
}
