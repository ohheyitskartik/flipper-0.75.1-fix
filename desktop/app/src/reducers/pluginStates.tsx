/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import type {Actions} from '.';
import {deconstructPluginKey} from '../utils/clientUtils';

export type State = {
  [pluginKey: string]: any;
};

export const pluginKey = (serial: string, pluginName: string): string => {
  return `${serial}#${pluginName}`;
};

export type Action =
  | {
      type: 'SET_PLUGIN_STATE';
      payload: {
        pluginKey: string;
        state: Object;
      };
    }
  | {
      type: 'CLEAR_CLIENT_PLUGINS_STATE';
      payload: {clientId: string; devicePlugins: Set<string>};
    };

export default function reducer(
  state: State | undefined = {},
  action: Actions,
): State {
  if (action.type === 'SET_PLUGIN_STATE') {
    const newPluginState = action.payload.state;
    if (newPluginState && newPluginState !== state[action.payload.pluginKey]) {
      return {
        ...state,
        [action.payload.pluginKey]: {
          ...state[action.payload.pluginKey],
          ...newPluginState,
        },
      };
    }
    return {...state};
  } else if (action.type === 'CLEAR_CLIENT_PLUGINS_STATE') {
    const {payload} = action;
    return Object.keys(state).reduce((newState: State, pluginKey) => {
      // Only add the pluginState, if its from a plugin other than the one that
      // was removed. pluginKeys are in the form of ${clientID}#${pluginID}.
      const plugin = deconstructPluginKey(pluginKey);
      const clientId = plugin.client;
      const pluginId = plugin.pluginName;
      if (
        clientId !== payload.clientId ||
        (pluginId && payload.devicePlugins.has(pluginId))
      ) {
        newState[pluginKey] = state[pluginKey];
      }
      return newState;
    }, {});
  } else {
    return state;
  }
}

export const setPluginState = (payload: {
  pluginKey: string;
  state: Object;
}): Action => ({
  type: 'SET_PLUGIN_STATE',
  payload,
});
