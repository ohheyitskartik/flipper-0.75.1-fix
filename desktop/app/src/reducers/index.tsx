/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {combineReducers, Dispatch} from 'redux';
import application, {
  State as ApplicationState,
  Action as ApplicationAction,
} from './application';
import connections, {
  State as DevicesState,
  Action as DevicesAction,
} from './connections';
import pluginStates, {
  State as PluginStatesState,
  Action as PluginStatesAction,
} from './pluginStates';
import pluginMessageQueue, {
  State as PluginMessageQueueState,
  Action as PluginMessageQueueAction,
} from './pluginMessageQueue';
import notifications, {
  State as NotificationsState,
  Action as NotificationsAction,
} from './notifications';
import plugins, {
  State as PluginsState,
  Action as PluginsAction,
} from './plugins';
import supportForm, {
  State as SupportFormState,
  Action as SupportFormAction,
} from './supportForm';
import settings, {
  Settings as SettingsState,
  Action as SettingsAction,
} from './settings';
import launcherSettings, {
  LauncherSettings as LauncherSettingsState,
  Action as LauncherSettingsAction,
} from './launcherSettings';
import pluginManager, {
  State as PluginManagerState,
  Action as PluginManagerAction,
} from './pluginManager';
import healthchecks, {
  Action as HealthcheckAction,
  State as HealthcheckState,
} from './healthchecks';
import pluginDownloads, {
  State as PluginDownloadsState,
  Action as PluginDownloadsAction,
} from './pluginDownloads';
import usageTracking, {
  Action as TrackingAction,
  State as TrackingState,
} from './usageTracking';
import user, {State as UserState, Action as UserAction} from './user';
import JsonFileStorage from '../utils/jsonFileReduxPersistStorage';
import LauncherSettingsStorage from '../utils/launcherSettingsStorage';
import {launcherConfigDir} from '../utils/launcher';
import os from 'os';
import {resolve} from 'path';
import xdg from 'xdg-basedir';
import {createTransform, persistReducer} from 'redux-persist';
import {PersistPartial} from 'redux-persist/es/persistReducer';

import {Store as ReduxStore, MiddlewareAPI as ReduxMiddlewareAPI} from 'redux';
import storage from 'redux-persist/lib/storage';
import {TransformConfig} from 'redux-persist/es/createTransform';

export type Actions =
  | ApplicationAction
  | DevicesAction
  | PluginStatesAction
  | PluginMessageQueueAction
  | NotificationsAction
  | PluginsAction
  | UserAction
  | SettingsAction
  | LauncherSettingsAction
  | SupportFormAction
  | PluginManagerAction
  | HealthcheckAction
  | TrackingAction
  | PluginDownloadsAction
  | {type: 'INIT'};

export type State = {
  application: ApplicationState;
  connections: DevicesState & PersistPartial;
  pluginStates: PluginStatesState;
  pluginMessageQueue: PluginMessageQueueState;
  notifications: NotificationsState & PersistPartial;
  plugins: PluginsState & PersistPartial;
  user: UserState & PersistPartial;
  settingsState: SettingsState & PersistPartial;
  launcherSettingsState: LauncherSettingsState & PersistPartial;
  supportForm: SupportFormState;
  pluginManager: PluginManagerState & PersistPartial;
  healthchecks: HealthcheckState & PersistPartial;
  usageTracking: TrackingState;
  pluginDownloads: PluginDownloadsState;
};

export type Store = ReduxStore<State, Actions>;
export type MiddlewareAPI = ReduxMiddlewareAPI<Dispatch<Actions>, State>;

const settingsStorage = new JsonFileStorage(
  resolve(
    ...(xdg.config ? [xdg.config] : [os.homedir(), '.config']),
    'flipper',
    'settings.json',
  ),
);

const setTransformer = (config: TransformConfig) =>
  createTransform(
    (set: Set<string>) => Array.from(set),
    (arrayString: string[]) => new Set(arrayString),
    config,
  );

const launcherSettingsStorage = new LauncherSettingsStorage(
  resolve(launcherConfigDir(), 'flipper-launcher.toml'),
);

export default combineReducers<State, Actions>({
  application,
  connections: persistReducer<DevicesState, Actions>(
    {
      key: 'connections',
      storage,
      whitelist: [
        'userPreferredDevice',
        'userPreferredPlugin',
        'userPreferredApp',
        'userStarredPlugins',
      ],
    },
    connections,
  ),
  pluginStates,
  pluginMessageQueue: pluginMessageQueue as any,
  notifications: persistReducer(
    {
      key: 'notifications',
      storage,
      whitelist: ['blacklistedPlugins', 'blacklistedCategories'],
    },
    notifications,
  ),
  plugins: persistReducer<PluginsState, Actions>(
    {
      key: 'plugins',
      storage,
      whitelist: ['marketplacePlugins'],
    },
    plugins,
  ),
  supportForm,
  pluginManager: persistReducer<PluginManagerState, Actions>(
    {
      key: 'pluginManager',
      storage,
      whitelist: ['uninstalledPlugins'],
      transforms: [setTransformer({whitelist: ['uninstalledPlugins']})],
    },
    pluginManager,
  ),
  user: persistReducer(
    {
      key: 'user',
      storage,
    },
    user,
  ),
  settingsState: persistReducer(
    {key: 'settings', storage: settingsStorage},
    settings,
  ),
  launcherSettingsState: persistReducer(
    {
      key: 'launcherSettings',
      storage: launcherSettingsStorage,
      serialize: false,
      // @ts-ignore: property is erroneously missing in redux-persist type definitions
      deserialize: false,
    },
    launcherSettings,
  ),
  healthchecks: persistReducer<HealthcheckState, Actions>(
    {
      key: 'healthchecks',
      storage,
      whitelist: ['acknowledgedProblems'],
    },
    healthchecks,
  ),
  usageTracking,
  pluginDownloads,
});
