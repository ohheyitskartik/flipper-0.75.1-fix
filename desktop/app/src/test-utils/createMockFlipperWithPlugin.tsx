/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {
  render,
  RenderResult,
  act as testingLibAct,
} from '@testing-library/react';
import {queries} from '@testing-library/dom';

import {
  selectPlugin,
  starPlugin,
  selectDevice,
  selectClient,
} from '../reducers/connections';
import BaseDevice from '../devices/BaseDevice';

import {rootReducer} from '../store';
import {Store} from '../reducers/index';
import Client, {ClientQuery} from '../Client';

import {buildClientId} from '../utils/clientUtils';
import {Logger} from '../fb-interfaces/Logger';
import {PluginDefinition} from '../plugin';
import {registerPlugins} from '../reducers/plugins';
import PluginContainer from '../PluginContainer';
import {getPluginKey, isDevicePluginDefinition} from '../utils/pluginUtils';
import {getInstance} from '../fb-stubs/Logger';
import {initializeFlipperLibImplementation} from '../utils/flipperLibImplementation';

export type MockFlipperResult = {
  client: Client;
  device: BaseDevice;
  store: Store;
  pluginKey: string;
  sendMessage(method: string, params: any, client?: Client): void;
  createDevice(serial: string): BaseDevice;
  createClient(
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ): Promise<Client>;
  logger: Logger;
  togglePlugin(plugin?: string): void;
};

type MockOptions = Partial<{
  /**
   * can be used to intercept outgoing calls. If it returns undefined
   * the base implementation will be used
   */
  onSend?: (pluginId: string, method: string, params?: object) => any;
  additionalPlugins?: PluginDefinition[];
  dontEnableAdditionalPlugins?: true;
  asBackgroundPlugin?: true;
}>;

export async function createMockFlipperWithPlugin(
  pluginClazz: PluginDefinition,
  options?: MockOptions,
): Promise<MockFlipperResult> {
  const store = createStore(rootReducer);
  const logger = getInstance();
  initializeFlipperLibImplementation(store, logger);
  store.dispatch(
    registerPlugins([pluginClazz, ...(options?.additionalPlugins ?? [])]),
  );

  function createDevice(serial: string): BaseDevice {
    const device = new BaseDevice(
      serial,
      'physical',
      'MockAndroidDevice',
      'Android',
    );
    store.dispatch({
      type: 'REGISTER_DEVICE',
      payload: device,
    });
    device.loadDevicePlugins(store.getState().plugins.devicePlugins);
    return device;
  }

  async function createClient(
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ): Promise<Client> {
    query = query ?? {
      app: name,
      os: 'Android',
      device: device.title,
      device_id: device.serial,
      sdk_version: 4,
    };
    const id = buildClientId({
      app: query.app,
      os: query.os,
      device: query.device,
      device_id: query.device_id,
    });

    const client = new Client(
      id,
      query,
      null, // create a stub connection to avoid this plugin to be archived?
      logger,
      store,
      [
        ...(isDevicePluginDefinition(pluginClazz) ? [] : [pluginClazz.id]),
        ...(options?.dontEnableAdditionalPlugins
          ? []
          : options?.additionalPlugins?.map((p) => p.id) ?? []),
      ],
      device,
    );

    client.rawCall = async (
      method: string,
      _fromPlugin: boolean,
      params: any,
    ): Promise<any> => {
      const intercepted = options?.onSend?.(method, params);
      if (intercepted !== undefined) {
        return intercepted;
      }
      switch (method) {
        case 'getPlugins':
          // assuming this plugin supports all plugins for now
          return {
            plugins: [...store.getState().plugins.clientPlugins.keys()],
          };
        case 'getBackgroundPlugins':
          return {plugins: options?.asBackgroundPlugin ? [pluginClazz.id] : []};
        default:
          throw new Error(
            `Test client doesn't support rawCall method '${method}'`,
          );
      }
    };
    client.rawSend = jest.fn();

    // enable the plugin
    if (
      !isDevicePluginDefinition(pluginClazz) &&
      !store
        .getState()
        .connections.userStarredPlugins[client.query.app]?.includes(
          pluginClazz.id,
        )
    ) {
      store.dispatch(
        starPlugin({
          plugin: pluginClazz,
          selectedApp: client.query.app,
        }),
      );
    }
    if (!options?.dontEnableAdditionalPlugins) {
      options?.additionalPlugins?.forEach((plugin) => {
        if (!isDevicePluginDefinition(plugin)) {
          store.dispatch(
            starPlugin({
              plugin,
              selectedApp: client.query.app,
            }),
          );
        }
      });
    }
    await client.init();

    // As convenience, by default we select the new client, star the plugin, and select it
    if (!skipRegister) {
      store.dispatch({
        type: 'NEW_CLIENT',
        payload: client,
      });
    }
    return client;
  }

  const device = createDevice('serial');
  const client = await createClient(device, 'TestApp');

  store.dispatch(selectDevice(device));
  store.dispatch(selectClient(client.id));

  store.dispatch(
    selectPlugin({
      selectedPlugin: pluginClazz.id,
      selectedApp: client.query.app,
      deepLinkPayload: null,
      selectedDevice: device,
    }),
  );

  return {
    client,
    device: device as any,
    store,
    sendMessage(method, params, actualClient = client) {
      actualClient.onMessage(
        JSON.stringify({
          method: 'execute',
          params: {
            api: pluginClazz.id,
            method,
            params,
          },
        }),
      );
    },
    createDevice,
    createClient,
    logger,
    pluginKey: getPluginKey(client.id, device, pluginClazz.id),
    togglePlugin(id?: string) {
      const plugin = id
        ? store.getState().plugins.clientPlugins.get(id)
        : pluginClazz;
      if (!plugin) {
        throw new Error('unknown plugin ' + id);
      }
      store.dispatch(
        starPlugin({
          plugin,
          selectedApp: client.query.app,
        }),
      );
    },
  };
}

type Renderer = RenderResult<typeof queries>;

export async function renderMockFlipperWithPlugin(
  pluginClazz: PluginDefinition,
  options?: MockOptions,
): Promise<
  MockFlipperResult & {
    renderer: Renderer;
    act: (cb: () => void) => void;
  }
> {
  const args = await createMockFlipperWithPlugin(pluginClazz, options);

  function selectTestPlugin(store: Store, client: Client) {
    store.dispatch(
      selectPlugin(
        isDevicePluginDefinition(pluginClazz)
          ? {
              selectedPlugin: pluginClazz.id,
              selectedApp: null,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            }
          : {
              selectedPlugin: pluginClazz.id,
              selectedApp: client.query.app,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            },
      ),
    );
  }

  selectTestPlugin(args.store, args.client);

  const renderer = render(
    <Provider store={args.store}>
      <PluginContainer logger={args.logger} />
    </Provider>,
  );

  return {
    ...args,
    renderer: renderer as any,
    act(cb) {
      testingLibAct(cb);
      args.client.flushMessageBuffer();
    },
  };
}
