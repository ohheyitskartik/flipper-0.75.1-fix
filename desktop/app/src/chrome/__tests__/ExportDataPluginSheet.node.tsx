/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import {create, act, ReactTestRenderer} from 'react-test-renderer';
import {Provider} from 'react-redux';
import ExportDataPluginSheet from '../ExportDataPluginSheet';
import {FlipperPlugin, FlipperDevicePlugin} from '../../plugin';
import {getExportablePlugins, getPluginKey} from '../../utils/pluginUtils';
import {createMockFlipperWithPlugin} from '../../test-utils/createMockFlipperWithPlugin';
import {setPluginState} from '../../reducers/pluginStates';

class TestPlugin extends FlipperPlugin<any, any, any> {
  static details = {
    title: 'TestPlugin',
    id: 'TestPlugin',
  } as any;
}
TestPlugin.title = 'TestPlugin';
TestPlugin.id = 'TestPlugin';
TestPlugin.defaultPersistedState = {msg: 'Test plugin'};

class TestDevicePlugin extends FlipperDevicePlugin<any, any, any> {
  static details = {
    title: 'TestDevicePlugin',
    id: 'TestDevicePlugin',
  } as any;

  static supportsDevice() {
    return true;
  }
}
TestDevicePlugin.title = 'TestDevicePlugin';
TestDevicePlugin.id = 'TestDevicePlugin';
TestDevicePlugin.defaultPersistedState = {msg: 'TestDevicePlugin'};

test('SettingsSheet snapshot with nothing enabled', async () => {
  let root: ReactTestRenderer;
  const {
    store,
    togglePlugin,
    client,
    device,
    pluginKey,
  } = await createMockFlipperWithPlugin(TestPlugin, {
    additionalPlugins: [TestDevicePlugin],
  });

  togglePlugin();

  store.dispatch(
    setPluginState({
      pluginKey,
      state: {test: '1'},
    }),
  );

  expect(getExportablePlugins(store.getState(), device, client)).toEqual([]);

  // makes device plugin visible
  store.dispatch(
    setPluginState({
      pluginKey: getPluginKey(undefined, device, 'TestDevicePlugin'),
      state: {test: '1'},
    }),
  );

  expect(getExportablePlugins(store.getState(), device, client)).toEqual([
    {
      id: 'TestDevicePlugin',
      label: 'TestDevicePlugin',
    },
  ]);

  await act(async () => {
    root = create(
      <Provider store={store}>
        <ExportDataPluginSheet onHide={() => {}} />
      </Provider>,
    );
  });

  expect(root!.toJSON()).toMatchSnapshot();
});

test('SettingsSheet snapshot with one plugin enabled', async () => {
  let root: ReactTestRenderer;
  const {store, device, client, pluginKey} = await createMockFlipperWithPlugin(
    TestPlugin,
    {
      additionalPlugins: [TestDevicePlugin],
    },
  );

  // enabled, but no data
  expect(getExportablePlugins(store.getState(), device, client)).toEqual([]);

  store.dispatch(
    setPluginState({
      pluginKey,
      state: {test: '1'},
    }),
  );
  store.dispatch(
    setPluginState({
      pluginKey: getPluginKey(undefined, device, 'TestDevicePlugin'),
      state: {test: '1'},
    }),
  );
  expect(getExportablePlugins(store.getState(), device, client)).toEqual([
    {
      id: 'TestDevicePlugin',
      label: 'TestDevicePlugin',
    },
    {
      id: 'TestPlugin',
      label: 'TestPlugin',
    },
  ]);

  await act(async () => {
    root = create(
      <Provider store={store}>
        <ExportDataPluginSheet onHide={() => {}} />
      </Provider>,
    );
  });

  expect(root!.toJSON()).toMatchSnapshot();
});
