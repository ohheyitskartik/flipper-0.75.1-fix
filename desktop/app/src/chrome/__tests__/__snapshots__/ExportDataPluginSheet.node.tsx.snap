// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsSheet snapshot with nothing enabled 1`] = `
<div
  className="css-p0wmbe-View-FlexBox-FlexColumn"
>
  <div
    className="css-1edwc8r-View-FlexBox-FlexColumn"
  >
    <div
      className="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
    >
      <span
        className="css-meh2qi-Text"
      >
        Select the plugins for which you want to export the data
      </span>
      <div
        className="css-1houjzq-View-FlexBox-FlexColumn"
      >
        <div
          className="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
        >
          <div
            className="css-auhar3-TooltipContainer e1abcqbd0"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-wospjg-View-FlexBox-FlexRow epz0qe20"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-xsnw23-Text e19o3fcp0"
                >
                  TestDevicePlugin
                </span>
                <div
                  className="css-t4wmtk-View-FlexBox-Spacer e13mj6h80"
                />
                <input
                  checked={false}
                  className="css-1pxrk7-CheckboxContainer e28aqfo0"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1p0wwd3-View"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1yqvjo0"
    >
      <div
        className="css-wospjg-View-FlexBox-FlexRow epz0qe20"
      >
        <div
          className="css-t4wmtk-View-FlexBox-Spacer e13mj6h80"
        />
        <div
          className="css-12n892b"
        >
          <button
            className="ant-btn ant-btn-default"
            onClick={[Function]}
            type="button"
          >
            <span>
              Close
            </span>
          </button>
        </div>
        <div
          className="css-auhar3-TooltipContainer e1abcqbd0"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="ant-btn ant-btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span>
              Submit
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SettingsSheet snapshot with one plugin enabled 1`] = `
<div
  className="css-p0wmbe-View-FlexBox-FlexColumn"
>
  <div
    className="css-1edwc8r-View-FlexBox-FlexColumn"
  >
    <div
      className="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
    >
      <span
        className="css-meh2qi-Text"
      >
        Select the plugins for which you want to export the data
      </span>
      <div
        className="css-1houjzq-View-FlexBox-FlexColumn"
      >
        <div
          className="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
        >
          <div
            className="css-auhar3-TooltipContainer e1abcqbd0"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-wospjg-View-FlexBox-FlexRow epz0qe20"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-xsnw23-Text e19o3fcp0"
                >
                  TestDevicePlugin
                </span>
                <div
                  className="css-t4wmtk-View-FlexBox-Spacer e13mj6h80"
                />
                <input
                  checked={false}
                  className="css-1pxrk7-CheckboxContainer e28aqfo0"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1p0wwd3-View"
            />
          </div>
        </div>
        <div
          className="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
        >
          <div
            className="css-auhar3-TooltipContainer e1abcqbd0"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="css-1jrm6r3"
            >
              <div
                className="css-wospjg-View-FlexBox-FlexRow epz0qe20"
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <span
                  className="css-xsnw23-Text e19o3fcp0"
                >
                  TestPlugin
                </span>
                <div
                  className="css-t4wmtk-View-FlexBox-Spacer e13mj6h80"
                />
                <input
                  checked={false}
                  className="css-1pxrk7-CheckboxContainer e28aqfo0"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
            <div
              className="css-1p0wwd3-View"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1yqvjo0"
    >
      <div
        className="css-wospjg-View-FlexBox-FlexRow epz0qe20"
      >
        <div
          className="css-t4wmtk-View-FlexBox-Spacer e13mj6h80"
        />
        <div
          className="css-12n892b"
        >
          <button
            className="ant-btn ant-btn-default"
            onClick={[Function]}
            type="button"
          >
            <span>
              Close
            </span>
          </button>
        </div>
        <div
          className="css-auhar3-TooltipContainer e1abcqbd0"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="ant-btn ant-btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span>
              Submit
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
